src/mds/MDSRank.h

```diff
@@ -151,6 +151,27 @@ class Finisher;
 class ScrubStack;
 class C_ExecAndReply;
 
+struct MDSMetaRequest {
+private:
+  int op;
+  CDentry *dentry;
+  ceph_tid_t tid;
+public:
+  explicit MDSMetaRequest(int o, CDentry *d, ceph_tid_t t) :
+    op(o), dentry(d), tid(t) {
+    dentry->get(CDentry::PIN_PURGING);
+    dentry->reintegration_reqid = tid;
+  }
+  ~MDSMetaRequest() {
+    dentry->reintegration_reqid = 0;
+    dentry->put(CDentry::PIN_PURGING);
+  }
```

Comment:
Looking at this again, I see now this is highly specific code for MDSRank. Why not just use a `MDSContext` that does this destructor bit.

Alternatively, make this class generic and sub-class in StrayManager.

Reply:
Possibly we can reuse the `MDSMetaRequest` in future for other cases. Currently this will only be used by the `reintegration/migration`. So I am also okay to move this to `StrayManager`.

Reply:
Improved it by switching to `struct StrayEvalRequest : public MDSMetaRequest {}`.